Client/Frontend/Browser/Playlist/PlaylistMediaInfo.swift

```diff
@@ -0,0 +1,509 @@
+                        @unknown default:
+                            assertionFailure("Unknown Switch Case for AVPlayerItemStatus")
+                        }
+                    })
+                }
+            })
+        }
+    }
+}
+
+extension PlaylistMediaInfo {
+    
+    func thumbnailForURL(_ url: String) -> UIImage? {
+        let sourceURL = URL(string: url)
+        let asset = AVAsset(url: sourceURL!)
+        let imageGenerator = AVAssetImageGenerator(asset: asset)
+        let time = CMTimeMakeWithSeconds(2, preferredTimescale: 1)
+        guard let imageRef = try? imageGenerator.copyCGImage(at: time, actualTime: nil) else {
+            return nil
+        }
+        return UIImage(cgImage: imageRef)
+    }
+}
+
+// A resource manager/downloader that is capable of handling HLS streams and content requests
+// This is used to determine if a resource is streamable (HLS) and to be able to request chunks of that stream.
+class MediaResourceManager: NSObject, AVAssetResourceLoaderDelegate {
+    
+    private var completion: ((Error?) -> Void)?
+    private var dataRequests = [AVAssetResourceLoadingRequest]()
+    private var contentInfoRequest: AVAssetResourceLoadingRequest?
+    private lazy var session = URLSession(configuration: .ephemeral, delegate: nil, delegateQueue: .main)
+    private var data = Data()
+    
+    init(_ completion: @escaping (Error?) -> Void) {
+        self.completion = completion
+    }
+    
+    func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForLoadingOfRequestedResource loadingRequest: AVAssetResourceLoadingRequest) -> Bool {
+        // Load the content information..
+        if let contentRequest = loadingRequest.contentInformationRequest {
+            guard let requestURL = loadingRequest.request.url else { return false }
+            guard MediaResourceManager.isShimmedURL(requestURL) else { return false }
+            let originalURL = MediaResourceManager.unShimURL(requestURL)
+            
+            // Request for byte-ranges..
+            let request: URLRequest = {
+                let offset = loadingRequest.dataRequest?.currentOffset ?? 0
+                let length = loadingRequest.dataRequest?.requestedLength ?? 0
+                
+                var request = loadingRequest.request
+                request.url = originalURL
+                request.addValue("bytes=\(offset)-\(max(1, length - 1))", forHTTPHeaderField: "Range")
+                return request
+            }()
+            
+            self.session.dataTask(with: request) { [weak self] data, response, error in
+                guard let self = self else { return }
+                guard let response = response as? HTTPURLResponse else {
+                    self.completion?("Invalid Response")
+                    return
+                }
+                
+                if response.statusCode >= 200 && response.statusCode <= 299 {
```

Comment:
I'd recommend checking for exact status codes, misconfigured servers sending invalid status codes like (207/215) can cause problems.

Reply:
Fixed.